UITableView data source

andersio opened the pull request:
Dependency: #25.

The API blueprint:

```swift
tableView.reactive.collection(
	.animation(insert: .automatic, delete: .automatic, reload: .automatic),
	.editable(editPredicate, commitAction),
	.movable(movePredicate, moveAction)
) <~ items.producer.map { item, cellFactory in
	let cell = cellFactory("ItemCell")
	cell.textLabel?.text = item
	return cell
}
```

The design attempts to present the entirety of table view sourcing as simple as mapping a collection to cells with the feel of FRP. Hopefully it should be applicable to other collection views (probably except NSTableView).

Instead of binding from a collection, its delta producer is used instead. This should be sufficiently flexible for layering and interoperating with, say, collection diffing signal/property operators.

Note: Support of sections, editable rows and moveable rows are not part of the PR.

### To-do

- [x] 1D collection delta binding.
- [x] Test coverage for 1D collection delta binding.
- [x] 2D collection delta binding.
- [ ] Test coverage for 2D collection delta binding.
- [ ] Reevaluate the use of protocols and constraints.
- [ ] Configurations (e.g. animations).

andersio commented on Sources/UITableView.swift:
```diff
@@ -0,0 +1,83 @@
+
+extension Reactive where Base: UITableView {
+	public var collection: BindingTarget<(UITableView) -> UITableViewDataSource> {
+		return BindingTarget(lifetime: lifetime) { [weak base] in
```
Ideally this would be the lifetime of the table view. But it would require pulling in ReactiveCocoa for all Darwin targets, or unless we split the project like how RxSwift and RxCocoa do.

RuiAAPeres replied:
Do you think it would make sense to move this (ReactiveCollections) into `ReactiveCocoa/ReactiveSwift`? And then on `ReactiveCocoa/ReactiveCocoa` create the `UITableView` binding?

andersio replied:
I expect it to happen when the API & implementation are mature. Collection _is_ essential.

That's said I thought we have the repo because we can quickly iterate in this repo from scratch, aren't we? There are too many things yet to be designed which may lead to breaking changes.

AnnKatF commented:
I've tried to use this by importing `Delta.swift`, `UITableView.swift` and `ReactiveCollections.swift` into my project and it crashed when the items in ReactiveArray were modified. 
The error log indicates that there is an inconsistency between the number of items in the array and the number of cells being updated.

andersio commented:
@AnnKatF Thanks for trying it. But I do have to remind you that it is a minefield without test coverage. Still writing tests. 😆

andersio committed:
Revamped delta types and protocols.
1. `Delta` is now position agnostic so as to be generalized for unordered or index-unstable collections, and it carries the actual elements being inserted and deleted, instead of the indices.
1. `IndexingDelta` is a delta protocol introduced for collections maintaining stable indexes, e.g. `Array`.
1. `ArrayDelta` is a delta type introduced for all integer indexing collections.

andersio committed:
New dependency: ReactiveCocoa.

andersio committed:
Introduce data source binding for UITableView.
